### DISCUSSION TITLE
Powering Pico Pi w/ WS2812B

### DISCUSSION THREAD

**Discussion by IdidothBawx** [micropython/micropython#10074] [Hardware & Peripherals]
Absolute newbie here...

Putting together a busy light for my office cubicle; 3d printed a miniature stoplight w/ lenses, inserted a strip of 10 ws2812b leds, push button for toggling red/yellow/green/off, code resides on the board as main.py, etc...

All connected to my Pi Pico and, when USB connected to my computer, functional. Button cycles through the defined choices (red/yellow/green/off) with absolutely no challenges.

Testing w/ an external power supply to the USB wasn't yielding anything, so I added a block of code to blink the onboard LED. Now, when powered externally, I can see that the main.py is running successfully because of the onboard LED. However, no lights on the 2812b strand, no toggle, nothing...

External power supply is 5v 2.5A -- I believe this should be enough. What [entirely obvious and forehead-smacking frustratingly simple] thing am I missing here? 

Appreciate all of you!

**Comment by jimmo** [comment]
@IdidothBawx can you provide more info? Perhaps a photo of the wiring. How are you powering the strip (which pin is it connected to)?

When you use external power, is it via the USB port, or via one of the V pins?

Do you have a multimeter? Checking the voltage on the + pin on the strip would be a good thing to try.

**Reply by IdidothBawx** [reply]
@jimmo I appreciate you for taking the time to entertain my questions!

Super simplistic project on my Pico--
WS2812b strip w/ 10 LEDs
- 5V connected to pin40 (VBUS)
- Ground connected to pin38
- DI connected to GP28 (pin34)

Single pole button for choice selection/rotation--
- One side to pin36 (3v3)
- One side to GP22 (pin29)

When using external power, it's via USB port. Multimeter showing 5v all the way down the led strip.

**Reply by peterhinch** [reply]
Have you defined GP22 with
```python
from machine import Pin
p = Pin(22, Pin.IN, Pin.PULL_DOWN)
```
?

**Reply by IdidothBawx** [reply]
Here's my current code; will try your suggestion and post the results. Apologies for not including this earlier...

```
import time
from picozero import Button
from neopixel import Neopixel
from machine import Pin, Timer
time.sleep(1.5)

# Influences onboard LED to indicate power to the Pico
led = Pin(25, Pin.OUT)
timer = Timer()
def blink(timer):
    led.toggle()
timer.init(freq=2.5, mode=Timer.PERIODIC, callback=blink)

# Main program // 
numpix = 10
pixels = Neopixel(numpix, 0, 28, "GRB")

yellow = (255, 100, 0)
green = (0, 255, 0)
red = (255, 0, 0)
blue = (0, 0, 255)
white = (255, 255, 255)
none = (0, 0, 0)

pixels.fill(none)
pixels.show()

button = Button(22, False)
option = 0

def choice():
    global option
    if option == 0:
        pixels.set_pixel_line(7, 10, red)
        pixels.brightness(50)
        pixels.show()
    elif option == 1:
        pixels.set_pixel_line(3, 5, yellow)
        pixels.set_pixel_line(7, 10, none)
        pixels.brightness(50)
        pixels.show()
    elif option == 2:
        pixels.set_pixel_line(0, 2, green)
        pixels.set_pixel_line(3, 5, none)
        pixels.brightness(50)
        pixels.show()
    elif option == 3:
        pixels.fill(none)
        pixels.show()
    
    if option == 3:
        option = 0
    else:
        option = option + 1

button.when_pressed = choice
```

**Reply by IdidothBawx** [reply]
I've also discovered that repeated button presses--nothing terribly fast--will generate an error. Nothing which appears to stop the program, only creates a delay:

```
Traceback (most recent call last):
  File "/lib/picozero/picozero.py", line 1533, in _pin_change
EventFailedScheduleQueueFull: Button (pin 22) - choice not run due to the micropython schedule being full
```

Line 1533 (i.e. str(self), callback...) in the library isn't telling me much -- could be that I've not yet learned what it means:

```
except RuntimeError as e:
                    if str(e) == "schedule queue full":
                        raise EventFailedScheduleQueueFull(
                            "{} - {} not run due to the micropython schedule being full".format(
                                str(self), callback_to_run.__name__))
                    else:
                        raise e
```

**Comment by peterhinch** [comment]
What is this module:
```python
from picozero import Button
```
My Pico W has no such module.

**Reply by fivdi** [reply]
> What is this module:
> 
> ```python
> from picozero import Button
> ```
> 
> My Pico W has no such module.

I would imagine it's this https://github.com/RaspberryPiFoundation/picozero

**Reply by peterhinch** [reply]
Thank you. Mercifully I'd been spared that monstrosity until now. It explains the Line 1533 error: the `picozero.py` module uses interrupts and `micropython.schedule` to debounce the pushbutton and it fails if button presses occur too fast. There are better ways to debounce a pushbutton.

As a general comment, `picozero.py` is the antithesis of the "Zen of micropython". Pulling in nearly 2K lines of code to interface a pushbutton is simply bizarre. Quite aside from the inefficiency, in my opinion huge monolithic files discourage learning.

**Comment by peterhinch** [comment]
@IdidothBawx Where did you get `neopixel.py`? It is evidently different from the official neopixel driver which is built in to your firmware.

**Comment by peterhinch** [comment]
The reason this doesn't work is that pin 40, the VBUS pin, only acts as an output when the Pico is powered from USB. The LED strip is therefore unpowered when you run it from an external supply on pin 39. Simply move the V+ pin on the LED strip to pin 39 and it will work.

You may be interested in this alternative approach which uses the official neopixel library and a much more minimal switch library. I only have an 8-LED Neopixel so the numbers are adapted.
```py
from neopixel import NeoPixel  # Official driver
from machine import Pin
import uasyncio as asyncio
from primitives import ESwitch

async def blink():  # Influences onboard LED to indicate power to the Pico
    led = Pin(25, Pin.OUT)
    while True:
        led(not led())
        await asyncio.sleep_ms(500)
    
# Main program // 
numpix = 8  # I only have 8

yellow = (255, 100, 0)
green = (0, 255, 0)
red = (255, 0, 0)
off = (0, 0, 0)

def line(np, start=0, end=numpix, color=off):
    for x in range(numpix):
        np[x] = color if start <= x < end else off
    np.write()

async def main():
    asyncio.create_task(blink())
    sw = ESwitch(Pin(22, Pin.IN, Pin.PULL_DOWN), lopen=0)
    pixels = NeoPixel(Pin(28, Pin.OUT), numpix)
    line(pixels)  # All off
    while True:
        await sw.close.wait()  # Wait for switch to close
        sw.close.clear()
        line(pixels, 5, 8, red)
        await sw.close.wait()  # Wait for next closure
        sw.close.clear()
        line(pixels, 3, 5, yellow)
        await sw.close.wait()
        sw.close.clear()
        line(pixels, 0, 3, green)
        await sw.close.wait()
        sw.close.clear()
        line(pixels)

asyncio.run(main())
```
If you want to try it you'll need to create a `primitives` directory on your board and copy `__init__.py` and `events.py` from [here](https://github.com/peterhinch/micropython-async/tree/master/v3/primitives). It doesn't throw exceptions if you operate the switch quickly.

**Comment by IdidothBawx** [comment]
I'll review once again when I'm back in my office.

That said, the challenge is related to EXCLUSIVELY to power over USB.

I can connect/power the unit from my computer and everything works
perfectly.

When connecting to a power supply (e.g. powerbank, outlet) via USB, the
main.py will not function appropriately. I can confirm power to the Pico
via the onboard LED blink (see my code above), but the button choice
selection fails to illuminate any of the ws2812b LEDs -- each 5v power
option I try (options that are not my computer) push enough amperage;
multimeter confirms power all the way through.

**Reply by peterhinch** [reply]
In which case I'm baffled.

**Comment by Lobo-T** [comment]
WS2812 is a 5V device, the Pico is a 3.3V device.
The DI pin on the WS2812 high level should be atleast 0.7*Vdd.
![image](https://user-images.githubusercontent.com/13045607/208911278-7e0c0a36-87c6-4203-866a-a52b1088d46c.png)

Which with exactly 5V is: 5*0.7 = 3.5V.  And the Pico is just giving it 3.3V.

I'm guessing that you computer is actually outputting a bit less than 5V, especially when powering the LED string.  Or maybe your external supply outputs a bit more.  Anyway, you are probably just out of spec.

You could try connecting the WS2812 strip to Vsys insted of Vbus.  Then you will get the voltage drop over D1.
![image](https://user-images.githubusercontent.com/13045607/208912600-aac390fe-6f48-4779-a161-5c0411fff079.png)

But the correct solution is probably converting the 3.3V output to 5V.  I use this circuit:
![image](https://user-images.githubusercontent.com/13045607/208913063-79a0266c-3b46-4253-8e91-65c905f6f16e.png)

(This is the same as the Philips i2c level shifter circuit.  Explained here https://assets.nexperia.com/documents/application-note/AN10441.pdf).
